### PR TITLE
Improving application's handling of new advisors (#139)

### LINKED ISSUE #139: Change redirect for users without cohorts and create advisors/<advisor> route at account creation
Related to #86 . Currently, when an advisor doesn't have a cohort (as is the case for many of them, particularly new users), they are redirected to /advisors. This has been confusing for many of them because it shows a list of advisors and they think that it is a list of people who have access to Student Explorer, and since their name isn't on it, they think they don't have access. Taking them to their own cohort page (with additional text on that page that will be added in another issue) should resolve this confusion. An additional reason to make this change is that /advisors is being removed from header navigation and redirects (in another issue) because it is not relevant to end users, so this change is necessary as a part of that.

Currently, the advisors/<advisor uniqname> route isn't created until an advisor has at least one student assigned to them in a cohort. This is why a redirect to /advisors is necessary right now--because there isn't an advisors/<advisor uniqname> page to direct to. However, this appears to be created with the assumption that only admins would not have a cohort, but this is no longer true. Thus, we will now want this route to exist as soon as an advisor's account is created. Then, we need to change the redirect so it doesn't redirect users to /advisors ever.

The redirect is in https://github.com/tl-its-umich-edu/student_explorer/blob/develop/seumich/views.py lines 162-169

Acceptance criteria: all users have an advisors/<advisor uniqname> route created at the time of account creation, regardless of whether or not they have ever had a cohort. When advisors navigate to https://studentexplorer.ai.umich.edu/ ('/') they land on the advisors/<advisor uniqname> page instead of /advisors.

### PR DESCRIPTION
**Pull request by ssciolla**
Inspired by past conversations with @mfldavidson and a solution suggested yesterday by @jonespm, this PR primarily makes changes to the `AdvisorView` and `IndexView` classes in order to improve the user experience of advisors new to Student Explorer. More details about the changes are provided below. This PR aims to resolve #139.

1) Previously, a `try/except` clause was used in `IndexView` to either redirect users to their "My Students" page or redirect them to the "All Advisors" page (if the particular user did not have a Mentor model instance with a matching username). I deleted the bulk of the clause so that all users are redirected to their "My Students" page, which uses the `/advisors/{username}` route pattern.

2) In the `AdvisorView` class, both the `get_queryset` and `get_context_data` methods were updated. Now `get_queryset` will try to find a `Mentor` model instance associated with the user's `username`. If the attempt succeeds, the application will return a QuerySet of students associated with the advisor. If the attempt fails, an empty list will be returned, ultimately resulting in an empty table on the front end. I also changed `get_context_data` so that (depending on whether a Mentor object was found and set to `self.mentor` in `get_queryset`) a context dictionary will be returned with either the name info from the Mentor object or the User object. I opted to use a `" ".join()" over concatenation to prepare advisors' full names, as that seems to more cleanly handle missing name values.

3) A small change was made to the `student_list_partial.html` template so that a `hr` tag (a shaded line break) always appears under the advisor's name, regardless of whether the advisor has any associated students. If the advisor has no students, an empty table is currently shown; this will be addressed in a future pull request addressing issue #140.

4) I modified the line spacing of the `urlpatterns` variable of `seumich/urls.py` to be more consistent. I also removed unnecessary imports to `django.conf.settings`.

### CODE REVIEW

**Comment by jonespm** [comment]
Wow solved 3 Codacy issues too!

File: seumich/views.py [hunk 1]
```diff
@@ -8,18 +8,18 @@
...
 from tracking.utils import UserLogPageViewMixin
 
 import operator
 import logging
 
-from django.conf import settings
 from decouple import config
 from functools import reduce
 
+User = get_user_model()
```
Comment (jonespm) [inline seumich/views.py]: You only use this once, you can probably just call it directly on line 98 as
get_user_model().objects.get

I'd at least move it out of the global module namespace.

File: seumich/views.py [hunk 2]
```diff
@@ -95,19 +95,27 @@ class AdvisorView(LoginRequiredMixin, UserLogPageViewMixin, ListView):
...
-                                        username=self.kwargs['advisor'])
-        student_list = self.mentor.students.order_by('last_name').distinct()
-        student_list = student_list.prefetch_related(
-            'studentclasssitestatus_set__status',
-            'studentclasssitestatus_set__class_site',
-            'cohorts')
+        self.mentor = None
+        try:
+            self.mentor = Mentor.objects.get(username=self.kwargs['advisor'])
+            student_list = self.mentor.students.order_by('last_name').distinct()
```
Comment (mfldavidson) [inline seumich/views.py]:
```suggestion
            student_list = self.mentor.students.filter(id>0).order_by('last_name').distinct()
```
May as well implement this now to get rid of "Not Happened Yet" as a student in the list

Comment (ssciolla) [reply seumich/views.py]: Okay, I can do this, though it is a separate issue. How are you planning on identifying the instances of Not Happened Yet in the future? It's nice to not have them show up, but then we don't know that something went wrong and that an advisor doesn't have a student they want in their cohort. I'm guessing you have a separate way of determining this?

Comment (mfldavidson) [reply seumich/views.py]: It is ok to leave out for now. It's not urgent or anything. I just figured it was such a small change to code we're already changing. I don't mind if you reject.

The utility of the appearance of "Not Happened Yet" is to make it very apparent that "something is wrong" which can cue me to run that program to figure out which uniqname(s) caused the error. However, I discovered it's much easier to filter BG_STDNT_CHRT_MNTR by STDNT_KEY<1 which will tell me which cohorts I need to look at.
image

Comment (ssciolla) [reply seumich/views.py]: I think you're right. It's a small change, and you have other ways to check for those values. I'm working on this now; I also want to check that there aren't other places with student QuerySets that need to be filtered.

File: seumich/views.py (hunk 1 above)
Comment (ssciolla) [reply seumich/views.py]: Done. See [this commit](https://github.com/tl-its-umich-edu/student_explorer/pull/161/commits/8d66ae1ba1c83b33642b68ecc2b8393396587da1).

File: seumich/views.py (hunk 2 above)
Comment (ssciolla) [reply seumich/views.py]: Done. See [this commit](https://github.com/tl-its-umich-edu/student_explorer/pull/161/commits/dfd11a2cd8c18310e69961fbbce71ad15e4684d8). We'll need to review this carefully in test, since there aren't zero or negative values for `Student.id` in the default data in my local database.

**Comment by ssciolla** [comment]
FYI, I found that many of the other views in `seumich/views.py` were using the `id__gte=0` field lookup, so I modified the one I just added for consistency. This isn't really a significant change, since there are no student ids with 0 in the relevant database tables.
